CR for dockumentrick

uriwin opened the pull request.

uriwin committed:
added targil 3

uri-win committed:
removed BufferInputStream.java and use lastByteRead instead.

uri-win committed:
change function name to getDataManipulatorInputStream

uri-win committed:
divided FileETL.java  into two class.

uri-win committed:
added class

noam-toledano commented on file_data_result.csv:
```diff
@@ -0,0 +1,2 @@
+uri, Winkowski, 155
+No/am, Toled/ano, 103
```
no need to commit test files

noam-toledano commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+public class Main {
+    public static void main(String[] args) throws IOException, ParseException {
```
this is a lot of logic in the `main` that your system's user just has to guess how to implement. make this simpler for an outside user by adding classes to take care of this program's flow, and add a `readme.md` file that describes the system for other users

noam-toledano commented on src/main/java/cli/Cli.java:
```diff
@@ -0,0 +1,19 @@
+
+    public Cli(String[] args, Options options) throws ParseException {
+        this.options = options;
+        this.parser = new DefaultParser();
```
dependencies should be injected from the outside instead of being created in the same place they are used

noam-toledano commented on src/main/java/cli/Cli.java:
```diff
@@ -0,0 +1,19 @@
+
+import org.apache.commons.cli.*;
+
+public class Cli {
```
this name is confusing. only reading the classes name does not clarify what the difference is between `Cli` and `CommandLine`

noam-toledano commented on src/main/java/cli/StringToEnumTypeConverter.java:
```diff
@@ -0,0 +1,32 @@
+import fileFilter.FilterType;
+import manipulateActions.ManipulateActionsType;
+
+public class StringToEnumTypeConverter {
```
SRP and OCP- why is this responsible for converting every type of string to an `enum`? 
according to the Open-Closed Principle, code should be open to extension (adding new classes) but closed to changes (touching an existing class)

noam-toledano commented on src/main/java/cli/StringToEnumTypeConverter.java:
```diff
@@ -0,0 +1,32 @@
+            case "enclose":
+                return ManipulateActionsType.StringEncloser;
+            default:
+                throw new IllegalArgumentException("No data manipulation action exists");
```
including the name of the manipulation given  in the error would make debugging and log-researching much easier

noam-toledano commented on src/main/java/cli/dockumentrick/DockumentrickCliParser.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Iterator;
+import java.util.List;
+
+public class DockumentrickCliParser {
```
this class is very tightly coupled to a `file` source and a `file` output. perhaps you should change this class's name to reflect that and make it implement an interface that would share methods with other types of input/output

noam-toledano commented on src/main/java/cli/dockumentrick/DockumentrickOptions.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+
+public class DockumentrickOptions {
```
coupling- if you wanted to add a new mode of selecting the options other than a CLI (such as a GUI or a rest service), all this info would need to be re-written (also, it would be difficult to look for it without already knowing where it is)

noam-toledano commented on src/main/java/cli/dockumentrick/ManipulatorActionParser.java:
```diff
@@ -0,0 +1,67 @@
+
+        while (iterator.hasNext()) {
+            Option option = iterator.next();
+            String argumentName = option.getLongOpt();
```
why is this class coupled to `commons-cli`?

noam-toledano commented on src/main/java/cli/dockumentrick/ManipulatorActionParser.java:
```diff
@@ -0,0 +1,67 @@
+        return dataManipulatorActions;
+    }
+
+    public ManipulateAction getDataManipulationAction(Option option, Iterator<Option> iterator) {
```
adding a new manipulation requires changes in many, many different places. according to the Open-Closed Principle, your code should be **open** to extensions (adding new features) but **closed** to modifications (changing existing code)

noam-toledano commented on src/main/java/cli/dockumentrick/ManipulatorActionParser.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Iterator;
+import java.util.List;
+
+public class ManipulatorActionParser {
```
SRP- this class has very intimate knowledge of the inner implementation of each manipulation. this is a very large responsibility, and it adds coupling, reduces readability, and make adding a new manipulation a very intense change

noam-toledano commented on src/main/java/fileETL/FileETL.java:
```diff
@@ -0,0 +1,25 @@
+    private InputStream inputStream;
+    private OutputStream outputStream;
+
+    public FileETL(InputStream inputStream, OutputStream outputStream) {
```
this whole class can be replaced with `IOUtils.copy`

uriwin replied:
IOUtils.copy method is for buffered copying between sources.
The inputStream that i created does not support buffered reading.

noam-toledano commented on src/main/java/fileFilter/FilterType.java:
```diff
@@ -0,0 +1,7 @@
+
+public enum FilterType {
+    ELEMENT,
+    COLUMN,
```
have more indicative names

noam-toledano commented on src/main/java/fileFilter/filters/CsvColumnFilter.java:
```diff
@@ -0,0 +1,28 @@
+    public CsvColumnFilter(int columnPositionToFilter) {
+        this.filterType = FilterType.COLUMN;
+        this.isFilterActivated = false;
+        this.currentColumnPosition = 1;
```
indexes begin at 0

noam-toledano commented on src/main/java/fileFilter/IFilter.java:
```diff
@@ -0,0 +1,10 @@
+package fileFilter;
+
+public interface IFilter {
```
this interface is too complex for an outside user. the user should only need to use one method for each action. this requires calling 2

noam-toledano commented on src/main/java/fileFilter/filters/CsvColumnFilter.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public void updateFilter(int fileByte) {
+        if (fileByte == ',')
```
what if the `,` is escaped?

uriwin replied:
in java ',' cannot be escaped. 
why is it necessary?

noam-toledano commented on src/main/java/fileFilter/filters/XmlElementFilter.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public String getStartingTag() {
+        return "<" + tagNameToFilter + ">";
```
if this isn't dynamic, why not create it once in the `ctor`?

noam-toledano commented on src/main/java/fileFilter/filters/XmlElementFilter.java:
```diff
@@ -0,0 +1,47 @@
+            if (tagData.equals(getClosingTag()))
+                isFilterActivated = false;
+            if (tagData.equals(getStartingTag()))
+                isFilterActivated = true;
```
this will only return `true` for the `>` at the end of the tag. find a way to return the correct value at the right time

noam-toledano commented on src/main/java/fileFilter/filters/XmlElementFilter.java:
```diff
@@ -0,0 +1,47 @@
+        }
+        else {
+            if ((char) fileByte == '<'){
+                tagData += (char) fileByte;
```
if you keep casting `fileByte` to `char` every time you use it, just do it once at the beginning

noam-toledano commented on src/main/java/fileFormat/formats/Csv.java:
```diff
@@ -0,0 +1,17 @@
+import fileFormat.AbstractFileFormat;
+import fileFormat.FileFormatType;
+
+public class Csv extends AbstractFileFormat {
```
needs a more indicative name- class names should be nouns that describe the purpose of the class. this isn't just "a CSV", make the name represent the purpose

noam-toledano commented on src/main/java/fileFormat/formats/Txt.java:
```diff
@@ -0,0 +1,13 @@
+import fileFormat.AbstractFileFormat;
+import fileFormat.FileFormatType;
+
+public class Txt extends AbstractFileFormat {
```
don't create an empty class just so you have one for each file type, it's fine for `txt` files to use the base class if its functionality suits it

noam-toledano commented on src/main/java/fileFormat/formats/Xml.java:
```diff
@@ -0,0 +1,33 @@
+    public boolean isInMiddleOfTag;
+
+    public Xml(FileFormatType fileFormatType) {
+        this.startOfTag = '<';
```
should be a `CONST`

noam-toledano commented on src/main/java/fileFormat/formats/Xml.java:
```diff
@@ -0,0 +1,33 @@
+import fileFormat.FileFormatType;
+
+public class Xml extends AbstractFileFormat {
+    public char startOfTag;
```
why is this public? do you want outside users to be able to set this to something unrelated and break the implementation?

noam-toledano commented on src/main/java/fileFormat/formats/Xml.java:
```diff
@@ -0,0 +1,33 @@
+    @Override
+    public boolean isByteRelatedToFileFormat(int data) {
+        if (isInMiddleOfTag) {
+            if ((char) data == endOfTag)
```
always add `{}`s to your `if` statements, it makes the code much more readable

noam-toledano commented on src/main/java/inputStream/DataManipulatorInputStream.java:
```diff
@@ -0,0 +1,93 @@
+        this.fileFormat = fileFormat;
+        this.dataManipulatorAction = dataManipulatorAction;
+        this.dataToReturnCount = this.dataToReturnPos = 0;
+        this.lastByteRead = -2;
```
why `-2`? don't include unexplained magic numbers in your code

noam-toledano commented on src/main/java/fileFilter/AbstractFilter.java:
```diff
@@ -0,0 +1,14 @@
+    public boolean isFilterActivated;
+    public FilterType filterType;
+
+    public boolean isFilterActivated() {
```
this name does not clearly communicate whether the filter allows changing the file or not. Does a return value of `true` mean that the filter has filtered the value, or that it allows the value to pass?

noam-toledano commented on src/main/java/fileFormat/AbstractFileFormat.java:
```diff
@@ -0,0 +1,9 @@
+package fileFormat;
+
+public abstract class AbstractFileFormat {
```
why can't this whole class be implemented as a `filter`?

noam-toledano commented on src/main/java/fileFormat/FileFormatFactory.java:
```diff
@@ -0,0 +1,20 @@
+            case XML:
+                return new Xml(fileFormatType);
+            default:
+                throw new IllegalArgumentException("Invalid file format type");
```
"unsupported file type" would be more indicative

noam-toledano commented on src/main/java/inputStream/DataManipulatorInputStream.java:
```diff
@@ -0,0 +1,93 @@
+    private AbstractFileFormat fileFormat;
+    private ManipulateAction dataManipulatorAction;
+    private byte[] dataToReturnBuffer;
+    private int dataToReturnPos;
```
don't save space on variable names

noam-toledano commented on src/main/java/inputStream/DataManipulatorInputStream.java:
```diff
@@ -0,0 +1,93 @@
+        this.in = inputStream;
+        this.fileFormat = fileFormat;
+        this.dataManipulatorAction = dataManipulatorAction;
+        this.dataToReturnCount = this.dataToReturnPos = 0;
```
separate these for readability

noam-toledano commented on src/main/java/inputStream/DataManipulatorInputStream.java:
```diff
@@ -0,0 +1,93 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+public class DataManipulatorInputStream extends FilterInputStream {
```
consider inheriting from `bufferedInputStream` instead

noam-toledano replied:
this class is very messy and convoluted. make the changes I stated and go over it again to make sure it is readable and simple. I had a lot of trouble understanding what was happening in this class

noam-toledano commented on src/main/java/inputStream/DataManipulatorInputStreamDecorator.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public InputStream getDataManipulatorInputStream() throws IOException {
+        InputStream inputStream = new FileInputStream(inputFilePath);
```
why is this coupled to a `file` implementation?

noam-toledano commented on src/main/java/manipulateActions/IManipulateAction.java:
```diff
@@ -0,0 +1,9 @@
+package manipulateActions;
+
+public interface IManipulateAction {
+    public String manipulateData(String fileChunkData);
```
1. why is this name coupled to `file`s?
2. why are you passing chunks of data using a string and not a proper collection (such as a `List`)?

uriwin replied:
is passing bytearray ok?

noam-toledano replied:
you should use an abstraction instead of an implementation, but yes

noam-toledano commented on src/main/java/manipulateActions/IManipulateAction.java:
```diff
@@ -0,0 +1,9 @@
+package manipulateActions;
+
+public interface IManipulateAction {
```
this interface is too complex. the user shouldn't need to know they need to call a specific method, then, based on the information that was returned to them, know whether to call another method

noam-toledano commented on src/main/java/manipulateActions/AbstractManipulateAction.java:
```diff
@@ -0,0 +1,14 @@
+    private ManipulateActionsType manipulateActionType;
+
+    public AbstractManipulateAction (ManipulateActionsType manipulateActionType){
+        this.manipulateActionType = manipulateActionType;
```
why does this class need to have this property?

noam-toledano commented on src/main/java/manipulateActions/EscapeCharacterAppender.java:
```diff
@@ -0,0 +1,25 @@
+        super(manipulateActionType);
+
+        this.specialChar = specialChar;
+        this.escapeCharacter = "/";
```
the escape character is backslash (`\`), not slash (`/`)

noam-toledano commented on src/main/java/manipulateActions/ManipulateAction.java:
```diff
@@ -0,0 +1,62 @@
+        this.manipulatorFilters.add(manipulatorFilter);
+    }
+
+    public boolean isAllFiltersActivated() {
```
should be `areAllFiltersActivated`

noam-toledano commented on src/main/java/manipulateActions/ManipulateAction.java:
```diff
@@ -0,0 +1,62 @@
+        return true;
+    }
+
+    public boolean isActionHaveFilters() {
```
should be `doesActionHaveFilters`

noam-toledano commented on src/main/java/manipulateActions/ManipulateAction.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public boolean isAllFiltersActivated() {
+        if (!isActionHaveFilters())
```
in the event that `manipulatorFilters` is empty, the `for` loop in the following lines would just not run any amount of times, resulting in a return value of `true`. this is unnecessary

noam-toledano commented on src/main/java/manipulateActions/StringEncloser.java:
```diff
@@ -0,0 +1,34 @@
+        if (currentData.length() < stringToEnclose.length()) {
+            if (stringToEnclose.charAt(currentData.length()) == (char) data) {
+                currentData += (char) data;
+                return true;
```
what happens if it's not clear yet? for example, I want to enclose the word `Noam` and I find the word `Nose`. why do `N` and `O` return true?

uriwin committed:
middle of cr

uriwin committed:
working on commandLine.

still need to work on fileFormat filter.

uriwin committed:
working on commandLine.

uriwin committed:
finished resolving CR comments on CommandLine.

uriwin committed:
finished working on CommandLine

uriwin committed:
removed FIleETL and use IOUtils.copy instead.

edit README.md, conatins now info on the program.

uriwin committed:
edit .gitignore to ignore txt,xml,csv files

uriwin committed:
removed test files

uriwin committed:
removed unnecessary code in ManipulatorInputStream.java.

changed name to IOption.java.

uriwin committed:
added in CsvValidator: isInQuotes.

uriwin commented:
reopen

noam-toledano commented on src/main/java/commandLine/clientArguments/ArgumentTypeValidator.java:
```diff
@@ -0,0 +1,55 @@
+package commandLine.clientArguments;
```
why is this in `commandLine`?

noam-toledano commented on src/main/java/commandLine/clientArguments/ArgumentTypeValidator.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            ManipulateActionsType manipulateActionType = stringToActionEnumConverter.convertActionNameToActionType(argumentName);
+            return true;
+        } catch (Exception e) {
```
why are you catching every single exception? only catch the relevant ones

noam-toledano commented on src/main/java/commandLine/clientArguments/ArgumentTypeValidator.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+    public boolean isArgumentConnectedToInputFileSource(String argumentName){
+        return argumentName.equals("inputFilePath");
```
no magic strings

noam-toledano commented on src/main/java/commandLine/clientArguments/ArgumentTypeValidator.java:
```diff
@@ -0,0 +1,55 @@
+            return false;
+        }
+    }
+    public boolean isArgumentConnectedToInputFileSource(String argumentName){
```
why is this class coupled to a `file` source?

noam-toledano commented on src/main/java/commandLine/convertStringToEnum/StringToOutputSourceEnumConverter.java:
```diff
@@ -0,0 +1,14 @@
+
+import commandLine.extracte.sourceExtractor.SourceType;
+
+public class StringToOutputSourceEnumConverter {
```
all these classes can be implemented with `ENUM.valueOf()`

noam-toledano commented on src/main/java/commandLine/clientArguments/IOption.java:
```diff
@@ -0,0 +1,7 @@
+
+import org.apache.commons.cli.Options;
+
+public interface IOption {
```
is this an `option`? if not, this is a bad name

noam-toledano commented on src/main/java/commandLine/clientArguments/CommandLineOptions.java:
```diff
@@ -1,31 +1,32 @@
         this.cliOptions.addOption(getCsvColumnFilter());
     }
 
     public Option getInputFilePathOption() {
```
why are these all `public`? the moment you make a method or property `public`, it becomes a part of your contract with the class's user.

noam-toledano commented on src/main/java/commandLine/clientArguments/CommandLineOptions.java:
```diff
@@ -1,31 +1,32 @@
+    public Option getOutputFilePathOption() {
+        return Option.builder().required().longOpt("outputFilePath").hasArg().build();
     }
     public Option getBaseConverterOption() { return Option.builder().longOpt("changeNumBase").hasArg().desc("(h/o/b)").build();}
```
many magic strings. if you wanted to add a different way to use this application (such as a GUI, or a REST service), you would need to copy and paste all this information again, which would make changing it a nightmare. put the data somewhere else, and have this class get it from there

noam-toledano commented on src/main/java/commandLine/extracte/manipulateActionsExtractor/ActionDTO.java:
```diff
@@ -0,0 +1,50 @@
+    private Map<String,String> actionFilters;
+
+    public ActionDTO(){
+        this.manipulateActionName = "";
```
why is this an empty string? this will just cause hard-to-track unwanted functionality if a user fails to set the action name

noam-toledano commented on src/main/java/commandLine/extracte/manipulateActionsExtractor/CommandLineActionsExtractor.java:
```diff
@@ -0,0 +1,63 @@
+                lastActionDTO = initializeActionDTO(argument, globalFilters);
+            }
+            else if (argumentTypeValidator.isArgumentConnectedToInputFileSource(argumentName)){
+                globalFilters.put(argumentName, FilenameUtils.getExtension(argument.getValue()).toUpperCase());
```
this class should be looking at the arguments through a higher level lense. It shouldn't need to care about small things like a specific filter, this should be part of a lower-level component's responsibility.

noam-toledano commented on src/main/java/commandLine/extracte/manipulateActionsExtractor/CommandLineActionsExtractor.java:
```diff
@@ -0,0 +1,63 @@
+        return actions;
+    }
+
+    public ActionDTO initializeActionDTO(Option argument, Map<String,String> globalFilters) {
```
why is this `public`?

noam-toledano commented on src/main/java/commandLine/extracte/sourceExtractor/SourceDTO.java:
```diff
@@ -0,0 +1,27 @@
+    private String sourceValue;
+
+    public SourceDTO(){
+        this.sourceValue = "";
```
don't add empty strings, it will only make it harder to debug when you forget to set this value

noam-toledano commented on src/main/java/commandLine/parse/ActionDTOParser.java:
```diff
@@ -0,0 +1,68 @@
+        switch (manipulateActionsType) {
+            case BaseConverter:
+                BaseTypeConverter baseTypeConverter = new BaseTypeConverter();
+                Integer baseToChange = baseTypeConverter.convertStringToInt(actionArguments.get(0));
```
This function should definitely not know about the internals of each class, instead, it should pass them the values and let them parse them however they like.

noam-toledano commented on src/main/java/fileFilter/filters/CsvColumnFilter.java:
```diff
@@ -1,28 +1,43 @@
+        if (data == '\"' || data == '\'')
+        {
+            isInQuotes = !isInQuotes;
+        }
```
this will create a bug if you have an opening `"` and a closing `'`.

noam-toledano commented on src/main/java/fileFilter/filters/LimitedLinesFilter.java:
```diff
@@ -1,24 +1,29 @@
+import status.Status;
 
-public class LimitedLinesFilter extends AbstractFilter {
+public class LimitedLinesFilter extends AbstractStatus {
```
is it a filter or a status? if this class's responsibility is to `filter` data based on a condition, it should be `filter`. if this class's responsibility is to signify a `status`, it should be named using the word `status`.

noam-toledano commented on src/main/java/fileFilter/filters/XmlElementFilter.java:
```diff
@@ -1,47 +1,61 @@
+            status = Status.MORE_DATA_NEEDED;
+        }
+    }
+    public boolean isEqualToEndOfTag(char data){
```
why is this public? make use `public` as little as possible

noam-toledano commented on src/main/java/fileFilter/filters/XmlElementFilter.java:
```diff
@@ -1,47 +1,61 @@
-                isInMiddleOfTag = true;
-            }
+        else if(tagData.equals(startingTagToFilter)){
+            status = Status.DATA_CAN_MANIPULATE;
```
why can you manipulate the tag itself?

noam-toledano commented on src/main/java/fileFormat/formats/CsvValidator.java:
```diff
@@ -0,0 +1,34 @@
+    public void updateStatus(char data)
+    {
+        updateIsInQuotes(data);
+        if ((data == ',' || data == '\n') && (!isInQuotes)){
```
no magic strings

noam-toledano commented on src/main/java/fileFormat/formats/XmlValidator.java:
```diff
@@ -0,0 +1,41 @@
+public class XmlValidator extends BaseFileValidator {
+    public boolean isInMiddleOfTag;
+
+    final static char START_OF_TAG = '<';
```
why `static`?

noam-toledano commented on src/main/java/fileFormat/formats/XmlValidator.java:
```diff
@@ -0,0 +1,41 @@
+
+        this.isInMiddleOfTag = false;
+
+        this.fileFormatType = fileFormatType;
```
doesn't this happen in `super`?

noam-toledano commented on src/main/java/inputStream/ManipulatorInputStream.java:
```diff
@@ -0,0 +1,84 @@
+            dataBuffer = fillDataBuffer((char) lastByteRead).getBytes();
+            dataBufferPos = 0;
+        }
+        return dataBuffer[dataBufferPos++];
```
don't return a `value++`, while it is functional, it makes the code less readable by forcing the reader to parse what happens when

noam-toledano commented on src/main/java/manipulateActions/ManipulateAction.java:
```diff
@@ -1,62 +1,64 @@
-        if (isActionHaveFilters()) {
-            for (IFilter manipulatorFilter : getManipulatorFilters()) {
-                manipulatorFilter.updateFilter(fileByte);
+    public boolean isFiltersStatusGood() {
```
you don't need both this and `isFiltersStatusBad`, choose one, then use `!` on the result

noam-toledano commented on src/main/java/manipulateActions/actions/BaseConverter.java:
```diff
@@ -0,0 +1,45 @@
+        return convertFromBaseToBase(data, this.currentBase, this.baseToChange);
+    }
+
+    public String convertFromBaseToBase(String data, int fromBase, int toBase) {
```
why is this public

noam-toledano commented on src/main/java/manipulateActions/actions/EscapeCharacterAppender.java:
```diff
@@ -0,0 +1,29 @@
+    public EscapeCharacterAppender(char specialChar) {
+        this.specialChar = specialChar;
+
+        this.escapeCharacter = "\\";
```
why not make this a `const`?

noam-toledano commented on src/main/java/inputStream/ManipulatorInputStream.java:
```diff
@@ -0,0 +1,84 @@
+
+    public String fillDataBuffer(char data) throws IOException {
+        String dataToFill = getDataToFill(data);
+        if (dataManipulatorAction.getStatus() == Status.DATA_CAN_MANIPULATE
+                && dataManipulatorAction.isFiltersStatusGood()) {
+            return dataManipulatorAction.manipulateDataAction(dataToFill);
```
why is this the `inputStream`'s responsibility to check this?
an API shouldn't require calling more than one method to perform an action if all it depends on is itself

noam-toledano reviewed (changes requested):
too many `public` methods. 
if you make a method `public` (or even `protected`), it becomes a part of your contract with the outside and cannot be changed. if a user doesn't need something, don't provide it

uriwin committed:
resolved some CR comments

uriwin committed:
ten comments left

uriwin committed:
last comment to resolve

uriwin committed:
finished resolving all CR comments

uriwin committed:
change AbstractStatus to AbstractStatusProvider.

fixed bug in limited lines filter.FilterFactory
changed name of DataManipulatorInputStreamDecorator to
ManipulatorInputStreamDecorator

noam-toledano commented on src/main/java/arguments/CommandLineArguments.java:
```diff
@@ -0,0 +1,61 @@
+        this.cliArguments.addOption(getLimitedLinesFilterOption());
+        this.cliArguments.addOption(getXmlElementFilterOption());
+        this.cliArguments.addOption(getCsvColumnFilter());
+    }
```
you could change this to a `foreach` loop on the enum, which would mean that adding new options to the enum will update this without needing to manually add it here as well

noam-toledano commented on src/main/java/commandLine/CommandLineCreator.java:
```diff
@@ -1,14 +1,14 @@
 import org.apache.commons.cli.ParseException;
 
 public class CommandLineCreator {
     public org.apache.commons.cli.CommandLine getCommandLine(String[] clientArguments) throws ParseException {
```
why not just import `CommandLine`?

noam-toledano commented on src/main/java/commandLine/CommandLineCreator.java:
```diff
@@ -1,14 +1,14 @@
 public class CommandLineCreator {
     public org.apache.commons.cli.CommandLine getCommandLine(String[] clientArguments) throws ParseException {
-        IOption options = new CommandLineOptions();
+        IArgumentsRequired argumentsRequired = new CommandLineArguments();
```
dependencies should be injected, not created in the same place they are used (DIP)

noam-toledano commented on src/main/java/commandLine/extract/destinationExtractor/CommandLineDestinationExtractor.java:
```diff
@@ -0,0 +1,29 @@
+                destinationDTO.setDestinationType(destinationType);
+                destinationDTO.setDestinationValue(option.getValue());
+                return destinationDTO;
+            } catch (Exception e) {}
```
don't swallow exceptions. this can make it very hard to debug issues with the program

noam-toledano commented on src/main/java/commandLine/parse/CommandLineParser.java:
```diff
@@ -19,37 +24,54 @@
+    }
+
+    private void addFilterBasedOnSource(SourceDTO inputSourceDTO, List<ActionDTO> actionDTOS){
+        if (inputSourceDTO.getSourceType() == SourceType.INPUT_FILE){
```
make this a `switch` statement to make it easier to add new input sources

noam-toledano commented on src/main/java/commandLine/parse/DestinationDTOParser.java:
```diff
@@ -0,0 +1,23 @@
+            case OUTPUT_FILE:
+                return new FileOutputStream(destinationDTO.getDestinationValue());
+            default:
+                throw new IllegalArgumentException("output source does not exists");
```
you should say it's `unsupported` instead

noam-toledano commented on src/main/java/fileFilter/filters/LimitedLinesFilter.java:
```diff
@@ -19,11 +19,11 @@ public void updateStatus(char data)
         if (data == '\n') {
```
no magic chars

uriwin committed:
resolved CR comments
